python/docs/source/user_guide/sql/python_udtf.rst

```diff
@@ -0,0 +1,140 @@
+..  Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+..    http://www.apache.org/licenses/LICENSE-2.0
+
+..  Unless required by applicable law or agreed to in writing,
+    software distributed under the License is distributed on an
+    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+    KIND, either express or implied.  See the License for the
+    specific language governing permissions and limitations
+    under the License.
+
+===========================================
+Python User-defined Table Functions (UDTFs)
+===========================================
+
+Spark 3.5 introduces a new type of user-defined fucntion: Python user-defined table functions (UDTFs),
+which take zero or more arguments and return a set of rows.
+
+Implementing a Python UDTF
+--------------------------
+
+.. currentmodule:: pyspark.sql.functions
+
+To implement a Python UDTF, you can implement this class:
+
+.. code-block:: python
+
+    class PythonUDTF:
+
+        def __init__(self) -> None:
+            """
+            Initialize the user-defined table function (UDTF).
+
+            This method is optional to implement and is called once when the UDTF is
+            instantiated. Use it to perform any initialization required for the UDTF.
+            """
+            ...
+
+        def eval(self, *args: Any) -> Iterator[Any]:
+            """"
+            Evaluate the function using the given input arguments.
```

Comment:
Can we also describe the mapping between provided scalar argument(s) and the attributes of any provided relation argument with the ordering of these `*args` here? For example, we could mention that each provided scalar argument with index i in 0...N gets assigned to `*args` index i here, with the exception of at most one input relation argument at provided index j, and its attributes are injected one-by-one into the `*args` starting at index j?

Reply:
I am thinking about this too, but I found it difficult to explain in words. The interface is the same as scalar UDFs so I think Spark users should be able to figure it out. I can provide more examples.

Reply:
👍 more examples should be helpful. Maybe we could also add:

```
The arguments provided to the UDTF call map to the values in this *args list
in order. Each provided scalar expression maps to exactly one value in this
*args list. Each provided TABLE argument of N columns maps to exactly N
values in this *args list, in order of the columns as they appear in the
table.
```